bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/TriggerHandlerCallbackImpl.java

```diff
@@ -43,21 +50,56 @@
 
     private final RuleEngineImpl re;
 
+    private final int threadPoolSize;
+
     protected TriggerHandlerCallbackImpl(RuleEngineImpl re, String ruleUID) {
         this.re = re;
         this.ruleUID = ruleUID;
-        executor = Executors.newSingleThreadExecutor(new NamedThreadFactory("rule-" + ruleUID));
+        this.threadPoolSize = ThreadPoolManager.getConfig(AUTOMATION_THREADPOOL_NAME);
+        if (threadPoolSize == -1) {
+            re.logger.debug("Firing rule {} using single thread.", ruleUID);
+            executor = Executors.newSingleThreadExecutor(new NamedThreadFactory("rule-" + ruleUID));
+        } else {
+            re.logger.debug("Firing rule {} using threadpool.", ruleUID);
+            executor = ThreadPoolManager.getPool(AUTOMATION_THREADPOOL_NAME);
+        }
     }
 
     @Override
     public void triggered(Trigger trigger, Map<String, ?> outputs) {
+        triggered(trigger, outputs, 0);
+    }
+
+    private void triggered(Trigger trigger, Map<String, ?> outputs, int itteration) {
+        RuleStatus ruleStatus = re.getRuleStatus(ruleUID);
         synchronized (this) {
             if (executor == null) {
                 return;
             }
-            future = executor.submit(new TriggerData(trigger, outputs));
+            if ((threadPoolSize == -1) || (RuleStatus.IDLE.equals(ruleStatus))) {
+                future = executor.submit(new TriggerData(trigger, outputs));
+                re.logger.debug("The trigger '{}' of rule '{}' is triggered. - Rule Status: '{}' - Rule Executed.",
+                        trigger.getId(), ruleUID, ruleStatus);
+            } else if ((RuleStatus.RUNNING.equals(ruleStatus)) && (itteration < maxItterations)) {
+                itteration++;
+                re.logger.debug("The trigger '{}' of rule '{}' is triggered.  Rule execution paused ({}/{})",
+                        trigger.getId(), ruleUID, itteration, maxItterations);
+                try {
+                    Thread.sleep(itterationWait);
```

Comment:
To control my lights I use proxy items and have a rule which handles the command to the proxy item and maps that to the command(s) to the real light item. So if I turn on all lights the same rule is going to want to run 12 times.

Am I right in thinking 12 threads are going to try and run the rule, but only one will succeed with the other 11 sleeping for 200ms? 200ms later one thread will run and the other 10 will sleep for 200ms. This continues until the final thread eventually gets a chance to run? If so this is going to result in a long delay before the final light turns on.

Reply:
I'm not sure I understand your example, but let me attempt to answer. This should only kick in when contention happens to prevent the same rule from executing if it's already executing. I'll do some tests but I believe this is per rule and not per rules file. We could very easily turn the timer down and make it iterate more before failing, my values we're initial spit ball guesses based on some very limited tests I did.  Also, this only happens in a threadpool environment (which seems to really only be applicable to systems at or under 2GB of RAM).  Everyone else above that shouldn't enable them. Single thread per rule environments (the current implementation in OH3) will just effectively queue up and fire as fast as possible.  So in your case, if you did have threadpools enabled, as long as you have a separate rule per light proxy item, they will all fire in parallel. This only applies to rules with multiple simultaneous trigger events.

Reply:
My example was one rule with 12 triggering items.
I'm running on an RPi3 and don't see any issues with OH, at what point do you feel you need to enable the thread pool?

Reply:
So the unfortunate answer is "there isn't a specific threshold".  I would point you to https://community.openhab.org/t/openhab-3-runs-out-of-memory-java-heap-space-errors-cpu-100-after-a-few-hours/110924 for a variety of users with a variety of setups (all seem to be Pi2/Pi3/small VMs) that are possibly being impacted here.  For some reason (which I haven't yet figured out, which I'm hoping this fixes), the Java VM runs out of memory and the whole system eventually goes sideways.  This was an example of memory usage from one user:

![0128b933a0b2c295a1954420b7b845a50ed411ad](https://user-images.githubusercontent.com/53797132/107882339-50915a00-6eae-11eb-8a57-a6cbe2d05959.png)

We have yet to figure out what happens between the red arrows but it's obvious that something starts to fail and isn't releasing.  I don't have this issue on my system so I'm somewhat guessing and hoping that this will fix it.  I do have some test jars that I'm going to post up later today for those users to load in hopes that they can at least confirm that it does fix the issue.  One thing of interesting note, it looks like most (but not all) of the users with these issues are using the new rules UI and not just text based rules.  

My hope/goal/theory is that there is some kind of large system event happening (some device updating and bursting a bunch of rules to fire at once) and there is a massive surge of threads being used by the rules engine.  By adding the threadpool as an option, we should be able to contain growth on these smaller systems into a pool (which is what OH2 used to do).